File: src/main/java/io/github/freshsupasulley/dbot/polls/Poll.java
package io.github.freshsupasulley.dbot.polls;

import java.time.Duration;
import java.util.function.Consumer;

import io.github.freshsupasulley.dbot.Main;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import net.dv8tion.jda.api.entities.messages.MessagePoll;
import net.dv8tion.jda.api.utils.messages.MessagePollData;

/**
 * Uses officially supported Discord polling functions. All of these kinds of polls last an hour.
 */
public abstract class Poll<T extends Poll<T>> {
	
	/** Defines the maximum allowed characters a poll can have before content is added. Not enforced */
	public static final int POLL_QUESTION_PREFIX = 50;
	
	private static final String YES_EMOJI = "U+2705", NO_EMOJI = "U+1f6ab";
	
	// Serialize everything
	private String question;
	private long messageID, startTime;
	private float ratio;
	private int minParticipation;
	private long votingCooldown;
	
	// Don't store
	private transient TextChannel channel;
	
	public Poll(float ratio, int minParticipation, long votingCooldown, String question, TextChannel channel)
	{
		this.ratio = ratio;
		this.minParticipation = minParticipation;
		this.votingCooldown = votingCooldown;
		this.question = question;
		this.channel = channel;
	}
	
	/**
	 * Checks if this poll equals another poll.
	 */
	@SuppressWarnings("unchecked")
	@Override
	public final boolean equals(Object obj)
	{
		if(this == obj)
			return true;
		if(obj == null || getClass() != obj.getClass())
			return false;
		
		return isDuplicate((T) obj);
	}
	
	/**
	 * Checks if the poll is a duplicate of another poll.
	 * 
	 * @param poll poll to check
	 * @return true if this poll is a duplicate, false otherwise
	 */
	public abstract boolean isDuplicate(T poll);
	
	public void firePoll(Consumer<Message> onSuccess, Consumer<Throwable> onFailure)
	{
		// Create message
		channel.sendMessagePoll(generatePoll()).queue(message ->
		{
			startTime = System.currentTimeMillis();
			messageID = message.getIdLong();
			onSuccess.accept(message);
		}, onFailure);
	}
	
	private MessagePollData generatePoll()
	{
		if(question.length() > MessagePoll.MAX_QUESTION_TEXT_LENGTH)
		{
			Main.log.error("{} has too long of a question", this);
		}
		
		return MessagePollData.builder(question.substring(0, Math.min(question.length(), MessagePoll.MAX_QUESTION_TEXT_LENGTH))).setDuration(getVoteTime()).addAnswer("Yes", Emoji.fromFormatted(YES_EMOJI)).addAnswer("No", Emoji.fromFormatted(NO_EMOJI)).build();
	}
	
	/**
	 * @return universal voting time for polls
	 */
	public Duration getVoteTime()
	{
		return Duration.ofDays(1);
	}
	
	private boolean passesPoll(int numYes, int numNo)
	{
		// Ignore if min participation wasn't met
		if(numYes + numNo <= minParticipation)
			return false;
		return numYes * 1f / (numYes + numNo) > ratio;
	}
	
	protected abstract void pollPassed(JDA jda);
	
	protected void pollFailed(JDA jda)
	{
	}
	
	/**
	 * Counts all votes and runs the associated poll action if passed.
	 * 
	 * <p>
	 * Ideally, this should check for naturalized citizens only but we're hiding the voting-booth channel from immigrants so this still should work.
	 * </p>
	 * 
	 * @param jda jda instance
	 */
	public void endPoll(JDA jda)
	{
		// Refresh channel
		channel = jda.getGuildById(Main.SERVER_ID).getTextChannelById(Main.VOTING_BOOTH);
		
		Message pollMessage = channel.retrieveMessageById(messageID).onErrorMap(t ->
		{
			Main.log.error("Failed to retrieve poll message", t);
			return null;
		}).complete();
		
		// Abandon if poll message couldn't get found
		if(pollMessage == null)
			return;
		
		MessagePoll poll = pollMessage.getPoll();
		// Due to the ordering when creating the poll, yes is always the first, no is always the second
		int numYes = poll.getAnswers().get(0).getVotes();
		int numNo = poll.getAnswers().get(1).getVotes();
		
		Main.log.info("To decide: Yes = " + numYes + ", No = " + numNo);
		
		// Delete voting message
		pollMessage.delete().queue();
		// Message afterMessage = null;
		
		// Check for ratio
		if(passesPoll(numYes, numNo))
		{
			Main.log.info("{} poll ({}) passed!", this, question);
			channel.sendMessage("**" + question + "** passed, with a Yes / No ratio of **" + numYes + "** / **" + numNo + "**!").complete();
			
			// Perform actions
			try
			{
				pollPassed(jda);
				Main.updateServerData();
			} catch(Exception e)
			{
				Main.log.error("Error running action during passed poll", e);
			}
		}
		else
		{
			// Fancy polling does this for us
			String failedMsg = "**" + question + "** failed to pass. Needs " + minParticipation + " voters and " + (int) (ratio * 100) + "% approval (Yes / No ratio: **" + numYes + "** / **" + numNo + "**)";
			channel.sendMessage(failedMsg).complete();
			Main.log.info(failedMsg);
			
			// Perform actions
			try
			{
				pollFailed(jda);
				Main.updateServerData();
			} catch(Exception e)
			{
				Main.log.error("Error running action during failed poll", e);
			}
		}
		
		// Matches Server's checkMessageForPollResult function
		// Idk how to check if its an Unknown Message
		// This is picked up by checkPolLResult function
		// afterMessage.delete().queueAfter(1, TimeUnit.HOURS, success -> {}, failure ->
		// {
		// DMain.log.error("Failed to delete poll after message", failure);
		// });
	}
	
	public long getMessageID()
	{
		return messageID;
	}
	
	public int getMinParticipants()
	{
		return minParticipation;
	}
	
	public long getVotingCooldown()
	{
		return votingCooldown;
	}
	
	public long getStartTime()
	{
		return startTime;
	}
}

File: src/polls/poll.ts
import { APIMessage, APIPoll, APIPollAnswer, PollLayoutType, RESTPostAPIChannelMessageJSONBody } from 'discord-api-types/v10';
import { InteractionResponseType } from 'discord-interactions';
import { globalState } from '..';
import { api } from '../utils';

export default abstract class BasePoll {
	type: string;

	// These get filled on firePoll()
	messageID!: string;
	startTime!: number;

	ratio: number;
	minParticipation: number;
	votingCooldown: number;
	// This is used in the result but not usually filled in the constructor
	question!: string;

	constructor(type: string, ratio: number, minParticipation: number, votingCooldown: number) {
		this.type = type;
		this.ratio = ratio;
		this.minParticipation = minParticipation;
		this.votingCooldown = votingCooldown;
	}

	public getExpiryTime(): number {
		return this.startTime + this.getVoteTime();
	}

	// Every poll (for now) can only have 1 day of voting time
	public getVoteTime(): number {
		return 8.64e7;
	}

	public getVotingCooldown(): number {
		return this.votingCooldown;
	}

	abstract isDuplicate(sample: BasePoll): boolean;
	abstract pollPassed(): Promise<any>;
	// Optional
	pollFailed(): Promise<any> {
		return Promise.resolve();
	}
	async firePoll() {
		// Create the poll object
		const poll: APIPoll = {
			allow_multiselect: false,
			expiry: new Date(Date.now() + this.getVoteTime()).toISOString(),
			question: {
				text: this.question,
			},
			layout_type: PollLayoutType.Default,
			answers: [
				{
					answer_id: 1,
					poll_media: {
						text: 'Yes',
						emoji: {
							id: null,
							name: '✅',
						},
					},
				},
				{
					answer_id: 2,
					poll_media: {
						text: 'No',
						emoji: {
							id: null,
							name: '🚫',
						},
					},
				},
			],
		};
		// Fire the poll
		return api(`channels/${globalState.serverData.votingBoothChannel}/messages`, {
			method: 'POST',
			body: {
				poll: poll,
			} as RESTPostAPIChannelMessageJSONBody,
		}).then((response: any) => {
			this.messageID = response.id;
			this.startTime = Date.now();
			// Add this poll to the serverData (and save it)
			globalState.serverData.polls.push(this);
			return {
				type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
				data: {
					content: `Poll added!`,
				},
			};
		});
	}

	private passesPoll(numYes: number, numNo: number): boolean {
		// Ignore if min participation wasn't met
		if (numYes + numNo <= this.minParticipation) return false;
		return (numYes * 1) / (numYes + numNo) > this.ratio;
	}

	/**
	 * Runs actions depending on the result of the poll. Does **NOT** delete the poll message!
	 * @param message original poll as {@link APIMessage}
	 */
	async endPoll(message: APIMessage) {
		const poll = message.poll!;
		console.log('The poll:', JSON.stringify(poll));
		// If no one votes, the answers can be undefined
		const numYes = poll.results?.answer_counts.find((count) => count.id === 1)?.count ?? 0;
		const numNo = poll.results?.answer_counts.find((count) => count.id === 2)?.count ?? 0;
		console.log(`To decide: Yes = ${numYes}, No = ${numNo}`);
		let response;
		// If we passed
		if (this.passesPoll(numYes, numNo)) {
			console.log('Poll passed');
			response = `**${this.question}** passed, with a Yes / No ratio of **${numYes}** / **${numNo}**!`;
			// Run whatever happens when it passes
			await this.pollPassed();
		} else {
			console.log('Poll failed');
			response = `**${this.question}** failed to pass. Needs ${this.minParticipation} voters and ${
				this.ratio * 100
			}% approval (Yes / No ratio: **${numYes}** / **${numNo}**)`;
			await this.pollFailed();
		}
		// Send the poll result in a garbage channel
		await api(`channels/${globalState.serverData.voteProposalChannel}/messages`, {
			method: 'POST',
			body: {
				content: response,
			},
		});
	}
}
